typing/ctype.ml

```diff
@@ -1971,32 +1971,53 @@ let enter_poly env univar_pairs t1 tl1 t2 tl2 f =
 
 let univar_pairs = ref []
 
-(* assumption: [ty] is fully generalized. *)
-let reify_univars ty =
-  let rec subst_univar scope vars ty =
+(**** Instantiate a generic type into a poly type ***)
+
+let polyfy env ty vars =
+  let rec subst_univars scope acc = function
+  | [] -> List.rev acc
+  | ty :: rest ->
+      let ty = repr ty in
+      match ty.desc with
+      | Tvar name when ty.level = generic_level ->
+          For_copy.save_desc scope ty ty.desc;
+          let t = newty (Tunivar name) in
+          ty.desc <- Tsubst t;
+          subst_univars scope (t :: acc) rest
+      | _ -> subst_univars scope acc rest
+  in
```

Comment:
This is a `List.filter_map`, right? (with copy-style side-effects)

Reply:
It is. In my defense, I think this code predates that function being in the stdlib.

Reply:
Could you use `List.filter_map` here (i.e. define `subst_univar` rather than `subst_univars`).
I think it clarifies the code.